### ISSUE TITLE
Any plans to add prompt caching for Anthropic?

### ISSUE THREAD

**Issue by tavurth** [karthink/gptel#365]
Why?

> When you send a request with Prompt Caching enabled:
> 
> The system checks if the prompt prefix is already cached from a recent query.
> If found, it uses the cached version, reducing processing time and costs.
> Otherwise, it processes the full prompt and caches the prefix for future use.
> 
> - Cache write tokens are 25% more expensive than base input tokens
> - Cache read tokens are 90% cheaper than base input tokens
> - Regular input and output tokens are priced at standard rates
> 
> This is especially useful for:
> 
> - Prompts with many examples
> - Large amounts of context or background information
> - Repetitive tasks with consistent instructions
> - Long multi-turn conversations
> 

Seems like it can be super useful for large codebases where only a small section needs to be updated each time. Can basically send it all over and it'll be cached each time.

https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching#how-prompt-caching-works

**Comment by CestDiego** [comment]
this would be perfect for having a similar feature to projects from Anthropic as well

**Comment by karthink** [comment]
Yup, will add it soon.

**Comment by axelknock** [comment]
The prompt caching API was recently simplified, so this should be easier to implement now:

> Quality-of-life upgrade for 
> [@AnthropicAI](https://x.com/AnthropicAI)
>  devs:
> 
> We've adjusted prompt caching so that you now only need to specify cache write points in your prompts - we'll automatically check for cache hits at previous positions. No more manual tracking of read locations needed.
[Source](https://x.com/alexalbert__/status/1879917906294870196)

**Comment by LittleHouse75** [comment]
Is prompt caching for Claude still on the roadmap?

I'm one of those folks whose conversations often begin 50-80k of text. Currently using TypingMind because it supports prompt caching, but would love to move my entire workflow to emacs :)

**Comment by karthink** [comment]
@LittleHouse75 Yeah I implemented this a while ago (basically after @axelknock's message above about the API simplification), but didn't get around to testing it.

I can push it now, but would like some input on how this should be used -- do you think ti makes sense for this to be a dynamic setting/user option, or set as a backend property?

#### Option 1

```elisp
(setq gptel-cache '(message tool system))    ;cache tool definitions, system prompt and conversation messages
```

#### Option 2

```elisp
(gptel-make-anthropic "Claude"
  :stream t
  :key "...."
  :cache '(message tool system))
```

**Comment by skyfield-ai** [comment]
I'd vote for backend - when would one not want to use it if a model offers it?

**Comment by lispy-ai** [comment]
Backend. Under what circumstances would one ever want to not use it if a model supports it? Unless there is such a circumstance it's not something one would change dynamically is it?

**Comment by karthink** [comment]
> Backend. Under what circumstances would one ever want to not use it if a model supports it? Unless there is such a circumstance it's not something one would change dynamically is it?

If you're making mostly one off requests as opposed to long conversations, you pay 25% extra for each request but rarely get a cache hit.  Let's hear what others have to say.

**Comment by lispy-ai** [comment]
That's a reason. I was unaware. 

So the longer the thread the more cost benefit it provides?

So this boils down to use case. Perhaps it could be set based on usage.

For example if one starts a thread in a gptel buffer, use it by default. If one uses a model for a one shot to do something, for example rewrite, don't enable it.

**Comment by karthink** [comment]
> So this boils down to use case. Perhaps it could be set based on usage.
>
> For example if one starts a thread in a gptel buffer, use it by default. If one uses a model for a one shot to do something, for example rewrite, don't enable it.

It can't be automated this way.  You might have a long system message (1000+ lines) that you use consistently but for one-off tasks.  In this case you'll want to cache the system message but not the messages array (i.e. conversation part of the prompt).  You can't cache based on the size of the message being sent because it might be a one-off query where you send a whole directory of files.  Only the user knows if an interaction is expected to be additive/repeated.

Since 0.9.7 gptel has been caching the prompt automatically whenever you include a media file (image/pdf etc).  I haven't exposed this capability to the user because I couldn't figure out how to provide the lever to control this, and lost track of it after a while.  Thus the current question of options 1 vs 2.

**Comment by lispy-ai** [comment]
In light of all this it seems clear it needs to be dynamic, easily accessed and persistently notified - reported in the status bar in GPTel buffers for example. 

The challenge for many users will likely be to decide when it is a benefit to enable it or not.

In a nutshell, when there is a lot of data that is included in the system or user prompt which does not change over repeated exchanges with the model.

By way of example for a common use case for me - multiple document analysis and code development - both have a lot of static content at each turn so they benefit from the cache.

My use is likely to want to default it on but to make it trivially easy to disable it as part of a one-shot (something I frequently do while working on multiple long running threads).

Context switching can be a pain in GPTel, especially as it relates to context as has been discussed in several other issues.

One thing that would make sense would be for cache enable to be tracked per GPTel buffer, just as the model and system prompt are now.

**Comment by LittleHouse75** [comment]
> [@LittleHouse75](https://github.com/LittleHouse75) Yeah I implemented this a while ago (basically after [@axelknock](https://github.com/axelknock)'s message above about the API simplification), but didn't get around to testing it.
> 
> I can push it now, but would like some input on how this should be used -- do you think it makes sense for this to be a dynamic setting/user option, or set as a backend property?
> 
> #### Option 1
> 
> (setq gptel-cache '(message tool system))    ;cache tool definitions, system prompt and conversation messages
> #### Option 2
> 
> (gptel-make-anthropic "Claude"
>   :stream t
>   :key "...."
>   :cache '(message tool system))

@karthink -- My personal vote is for dynamic setting / user option, with it disabled by default. I think those who need it generally know they need it. Thanks so much for diving back into this!

**Comment by karthink** [comment]
I've added prompt caching to the Anthropic backend, please update and test.

It's difficult to tell if it's working.  Here's how to test it:
1. Set `gptel-cache` (see doc for allowed values)
2. Turn on logging: `(setq gptel-log-level 'info)`
3. Use any Claude model in a conversation-type scenario.
4. Search the `*gptel-log*` buffer for `"cache_creation_input_tokens"` and `"cache_read_input_tokens"`, and check if these numbers make sense.

**Comment by LittleHouse75** [comment]
This is huge for me. Thank you!

**Comment by karthink** [comment]
@LittleHouse75 please test that it's working!

**Comment by LittleHouse75** [comment]
@karthink Absolutely, I will test this evening after work. Really looking forward to it!

**Comment by LittleHouse75** [comment]
> [@LittleHouse75](https://github.com/LittleHouse75) please test that it's working!
> […](#)

So yes - Claude prompt caching seems to work beautifully.

Below are excerpts from the logs showing prompt caching triggering once the context size hits 1024, with much reduced input tokens thereafter.

This is awesome.

Do you have a patreon (or similar) where I could send you a coffee for your efforts?

```
data: {"type":"message_start","message":{"id":"msg_01AmyQ7K2pSggxyhLFVQajDL","type":"message","role":"assistant","model":"claude-3-7-sonnet-20250219","content":[],"stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":342,"cache_creation_input_tokens":0,"cache_read_input_tokens":0,"output_tokens":1}}               }

data: {"type":"message_start","message":{"id":"msg_017ZWk8dcyBQrrQ8o5CQ6Mzx","type":"message","role":"assistant","model":"claude-3-7-sonnet-20250219","content":[],"stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":534,"cache_creation_input_tokens":0,"cache_read_input_tokens":0,"output_tokens":1}}            }

data: {"type":"message_start","message":{"id":"msg_01NUo39b7DxfMWTC2aC9owMb","type":"message","role":"assistant","model":"claude-3-7-sonnet-20250219","content":[],"stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":813,"cache_creation_input_tokens":0,"cache_read_input_tokens":0,"output_tokens":1}}    }

data: {"type":"message_start","message":{"id":"msg_01JXjPVUURfSH6ADyzAXTFHd","type":"message","role":"assistant","model":"claude-3-7-sonnet-20250219","content":[],"stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":4,"cache_creation_input_tokens":1039,"cache_read_input_tokens":0,"output_tokens":1}}    }

data: {"type":"message_start","message":{"id":"msg_01N2jxgkVP2mnJstF7BuwKFJ","type":"message","role":"assistant","model":"claude-3-7-sonnet-20250219","content":[],"stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":4,"cache_creation_input_tokens":272,"cache_read_input_tokens":1039,"output_tokens":1}} }

data: {"type":"message_start","message":{"id":"msg_01Gk7yAvdfq8BfxV8uf85k8J","type":"message","role":"assistant","model":"claude-3-7-sonnet-20250219","content":[],"stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":4,"cache_creation_input_tokens":353,"cache_read_input_tokens":1311,"output_tokens":1}}      }

data: {"type":"message_start","message":{"id":"msg_01N6pw3H3BXb9rAXDsqVnykL","type":"message","role":"assistant","model":"claude-3-7-sonnet-20250219","content":[],"stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":4,"cache_creation_input_tokens":242,"cache_read_input_tokens":1664,"output_tokens":1}} }

```

**Comment by lispy-ai** [comment]
@karthink it appears that d9ce451 (incorrectly) references #355 and not this issue #365 to which that commit actually applies.
